### PR TITLE
EKF2: improve ring buffer sizing and dynamically allocate

### PR DESCRIPTION
**Pull request by dagar**
##### PR (defaults)
``` Console
DEBUG 973048     EKF max time delay 165.0 ms, OBS length 11

IMU average dt: 0.005024 seconds
IMU buffer: 11 (536 Bytes)
minimum observation interval 10048 us
gps buffer: 5/11 (536 Bytes)
mag buffer: 5/11 (272 Bytes)
baro buffer: 1/11 (184 Bytes)
output buffer: 11/11 (536 Bytes)
output vert buffer: 11/11 (272 Bytes)
```

##### PR (EKF2_AID_MASK 0)
``` Console
DEBUG 955316     EKF max time delay 75.0 ms, OBS length 5

IMU average dt: 0.005023 seconds
IMU buffer: 5 (248 Bytes)
minimum observation interval 10047 us
gps buffer: 3/5 (248 Bytes)
mag buffer: 1/5 (128 Bytes)
baro buffer: 1/5 (88 Bytes)
output buffer: 5/5 (248 Bytes)
output vert buffer: 5/5 (128 Bytes)

```

##### PR (SENS_MAG_RATE 15)
``` Console
DEBUG 991358     EKF max time delay 165.0 ms, OBS length 11

IMU average dt: 0.005024 seconds
IMU buffer: 11 (536 Bytes)
minimum observation interval 10048 us
gps buffer: 5/11 (536 Bytes)
mag buffer: 1/11 (272 Bytes)
baro buffer: 0/11 (184 Bytes)
output buffer: 11/11 (536 Bytes)
output vert buffer: 11/11 (272 Bytes)
```

##### PR (SENS_MAG_RATE 15, EKF2_AID_MASK 0)
``` Console
DEBUG 978155     EKF max time delay 99.0 ms, OBS length 7

IMU average dt: 0.005023 seconds
IMU buffer: 7 (344 Bytes)
minimum observation interval 10048 us
gps buffer: 3/7 (344 Bytes)
mag buffer: 0/7 (176 Bytes)
baro buffer: 0/7 (120 Bytes)
output buffer: 7/7 (344 Bytes)
output vert buffer: 7/7 (176 Bytes)
```

##### master (defaults)
``` Console
imu buffer: 12 (584 Bytes)
gps buffer: 9 (440 Bytes)
mag buffer: 9 (224 Bytes)
baro buffer: 9 (152 Bytes)
range buffer: 1 (24 Bytes)
airspeed buffer: 1 (24 Bytes)
flow buffer: 1 (48 Bytes)
vision buffer: 1 (112 Bytes)
output buffer: 12 (584 Bytes)
output vert buffer: 12 (296 Bytes)
drag buffer: 1 (24 Bytes)
```

##### master (EKF2_AID_MASK 0)
``` Console
imu buffer: 2 (104 Bytes)
gps buffer: 1 (56 Bytes)
mag buffer: 1 (32 Bytes)
baro buffer: 1 (24 Bytes)
range buffer: 1 (24 Bytes)
airspeed buffer: 1 (24 Bytes)
flow buffer: 1 (48 Bytes)
vision buffer: 1 (112 Bytes)
output buffer: 2 (104 Bytes)
output vert buffer: 2 (56 Bytes)
drag buffer: 1 (24 Bytes)
```

### CODE REVIEW

File: src/modules/ekf2/EKF/estimator_interface.cpp [hunk 1]
```diff
@@ -202,67 +206,61 @@ void EstimatorInterface::setGpsData(const gps_message &gps)
...
 	if ((baro_sample.time_us - _time_last_baro) > _min_obs_interval_us) {
 		_time_last_baro = baro_sample.time_us;
 
-		const float baro_alt_avg = _baro_alt_sum / (float)_baro_sample_count;
-
 		baroSample baro_sample_new;
-		baro_sample_new.hgt = compensateBaroForDynamicPressure(baro_alt_avg);
+		baro_sample_new.hgt = compensateBaroForDynamicPressure(baro_sample.hgt);
 
 		// Use the time in the middle of the downsampling interval for the sample
```
Comment (bresch) [inline src/modules/ekf2/EKF/estimator_interface.cpp:240]:
```suggestion
```

File: src/modules/ekf2/EKF/control.cpp [hunk 2]
```diff
@@ -85,47 +85,59 @@ void Ekf::controlFusionModes()
...
 
-	// check for intermittent data (before pop_first_older_than)
-	const baroSample &baro_init = _baro_buffer.get_newest();
-	_baro_hgt_faulty = !isRecent(baro_init.time_us, 2 * BARO_MAX_INTERVAL);
+	if (_baro_buffer) {
+		// check for intermittent data (before pop_first_older_than)
+		const baroSample &baro_init = _baro_buffer->get_newest();
+		_baro_hgt_faulty = !isRecent(baro_init.time_us, 2 * BARO_MAX_INTERVAL);
+
+		_delta_time_baro_us = _baro_sample_delayed.time_us;
```
[the remark this answers is not in the document]
Comment (dagar) [reply src/modules/ekf2/EKF/control.cpp:93]: Not added here, but yes I don't like it either. I'll fix it.

File: src/modules/ekf2/EKF/estimator_interface.cpp [hunk 3]
```diff
@@ -104,70 +117,61 @@ bool EstimatorInterface::checkIfVehicleAtRest(float dt, const imuSample &imu)
...
+			_gps_buffer = nullptr;
 			printBufferAllocationFailed("GPS");
 			return;
 		}
 	}
 
-	// limit data rate to prevent data being lost
-	const bool need_gps = (_params.fusion_mode & MASK_USE_GPS) || (_params.vdist_sensor_type == VDIST_SENSOR_GPS);
-
 	// TODO: remove checks that are not timing related
```
[the remark this answers is not in the document]
Comment (dagar) [reply src/modules/ekf2/EKF/estimator_interface.cpp]: Done.

File: src/modules/ekf2/EKF/ekf.cpp [hunk 4]
```diff
@@ -519,9 +522,9 @@ Quatf Ekf::calculate_quaternion() const
 {
 	// Correct delta angle data for bias errors using bias state estimates from the EKF and also apply
 	// corrections required to track the EKF quaternion states
-	const Vector3f delta_angle{_newest_high_rate_imu_sample.delta_ang - _state.delta_ang_bias * (_dt_imu_avg / _dt_ekf_avg) + _delta_angle_corr};
+	const Vector3f delta_angle{_newest_high_rate_imu_sample.delta_ang - _state.delta_ang_bias *(_dt_imu_avg / _dt_ekf_avg) + _delta_angle_corr};
 
 	// increment the quaternions using the corrected delta angle vector
 	// the quaternions must always be normalised after modification
-	return Quatf{_output_new.quat_nominal * AxisAnglef{delta_angle}}.unit();
+	return Quatf{_output_new.quat_nominal *AxisAnglef{delta_angle}}.unit();
```
Comment (bresch) [inline src/modules/ekf2/EKF/ekf.cpp:529]: no

Comment (dagar) [reply src/modules/ekf2/EKF/ekf.cpp:529]: It's astyle.

**Review by bresch — APPROVED** [review]

**Pull request merged**